dvc/remote/local/__init__.py

```diff
@@ -581,3 +586,52 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if not self.changed(
+            path_info, {self.PARAM_CHECKSUM: checksum}
+        ) and self._is_same_link_as_cache(path_info):
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
```

Comment:
I don't think we need this message, as self.changed already prints it.

```suggestion
```

Reply:
And in that case, having this wrapper function seems a bit excessive :slightly_smiling_face: You could totally replace 
```
       if self._needs_checkout(path_info, checksum):
```
in _checkout_file with
```
       if not self._link_type_matches(path_info) or self.changed(path_info, checksum):
           ...
```
and have it more readable, since it describes it in one statement. Also, note that I've put link type detection first, since it should be faster than `self.changed()`.

Reply:
done